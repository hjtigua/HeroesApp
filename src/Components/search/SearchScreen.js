import React, { useMemo } from "react";
import { useLocation } from "react-router";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../heroes/HeroCard";
import queryString from "query-string";
import { getHeroesByName } from "../../selectors/getHeroesByName";

export const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  const [formValues, handleInputChangue] = useForm({
    searchString: q,
  });

  const { searchString } = formValues;
  const handleSubmit = (e) => {
    e.preventDefault();
    history.push(`?q=${searchString}`);
  };

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  return (
    <div>
      <h1>SeachScreen</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              type="text"
              placeholder="Find your Hero"
              className="form-control"
              name="searchString"
              value={searchString}
              autoComplete="off"
              onChange={handleInputChangue}
            />
            <button type="submit" className="btn btn-primary mt-2">
              Search
            </button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {q === "" && <div className="alert alert-info">Search a hero</div>}
          {q !== "" && heroesFiltered.length === 0 && (
            <div className="alert alert-danger">Hero no found</div>
          )}

          {heroesFiltered.map((hero) => {
            return <HeroCard key={hero.id} hero={hero} />;
          })}
        </div>
      </div>
    </div>
  );
};
