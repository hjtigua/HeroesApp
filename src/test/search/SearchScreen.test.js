import { mount } from "enzyme";
import React from "react";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../Components/search/SearchScreen";

/*
Aqui se reenderiza el "SearchScreen" desde el Router unicamente xq dentro utiliza hooks
para poder trabajar con rutas

*/

describe("Pruebas en SearchScren", () => {
  const wrapper = mount(
    <MemoryRouter initialEntries={["/search"]}>
      <Route path="/search" component={SearchScreen}></Route>
    </MemoryRouter>
  );

  test("debe de mostrarse correctamente con valores por defecto", () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toMatchSnapshot(
      "Search a hero"
    );
  });

  test("debe de mostrar a batman y el input con el valor del query string", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route path="/search" component={SearchScreen}></Route>
      </MemoryRouter>
    );

    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de mostrar un error sino se encuentra el hero", () => {
    const wrapper = mount(
      <MemoryRouter
        initialEntries={["/search?q=ysiehdemorirnoquierosabernadadenadie"]}
      >
        <Route path="/search" component={SearchScreen}></Route>
      </MemoryRouter>
    );

    expect(wrapper.find(".alert-danger").text().trim()).toBe("Hero no found");
  });

  test("debe de llamar el push del history", () => {
    const history = {
      push: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter
        initialEntries={["/search?q=ysiehdemorirnoquierosabernadadenadie"]}
      >
        <Route
          path="/search"
          component={() => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );

    wrapper.find("input").simulate("change", {
      target: {
        name: "searchString",
        value: "batman",
      },
    });

    wrapper.find("form").prop("onSubmit")({
      preventDefault() {},
    });

    expect(history.push).toHaveBeenCalledWith(`?q=batman`);
  });
});
