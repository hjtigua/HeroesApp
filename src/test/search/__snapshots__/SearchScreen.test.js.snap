// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScren debe de mostrar a batman y el input con el valor del query string 1`] = `
<div>
  <h1>
    SeachScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchString"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary mt-2"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card"
        >
          <img
            alt="..."
            className="card-img-top"
            src="./assets/heroes/dc-batman.jpg"
          />
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
              Batman
            </h5>
            <p
              className="card-text"
            >
              Bruce Wayne
               
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Ver MÃ¡s
              </a>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearchScren debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SeachScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchString"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary mt-2"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en SearchScren debe de mostrarse correctamente con valores por defecto: Search a hero 1`] = `"Search a hero"`;
